protocol/__init__.py

```diff
@@ -0,0 +1,136 @@
+            resolve = send(
+                self.w3, 
+                self.datatrust_key, 
+                self.datatrust.resolve_registration(
+                    self.w3.sha3(text=self.datatrust_host), {'from': self.datatrust_wallet}
+                    )
+                )
+            log.info(f'Resolved, transaction id: {resolve}')
+        else:
+            # backend not set, or is set to a different host
+            datatrust_url = call(self.datatrust.get_backend_url())
+            if datatrust_url == self.datatrust_host:
+                log.info('Server has been registered as datatrust, but not voted in')
+            else:
+                log.info('No backend or different host set. Resolving prior registrations and Submitting this one for voting')
+                datatrust_hash = self.w3.sha3(text=self.datatrust_host)
+                #TODO: Get the hash of any pending candidates and resolve before registration
+                resolve = send(
+                    self.w3, 
+                    self.datatrust_key, 
+                    self.datatrust.resolve_registration(
+                        self.w3.sha3(text=self.datatrust_host), {'from': self.datatrust_wallet}
+                        )
+                    )
+                log.info(f'Resolved any prior registration, transaction id: {resolve}')
+                register = self.register_host()
+                log.info(f'Datatrust has been registered, transaction id: {register}')
+
+    def wait_for_vote(self):
+        """
+        Check if this backend is registered as the datatrust
+        """
+        is_host = False
+        while is_host == False:
+            backend = call(self.datatrust.get_backend_url())
+            if backend != self.datatrust_host:
+                print('backend not voted in yet, waiting for votes...')
+                sleep(30)
+            else:
+                is_host = True
+        return True
+
+    def register_host(self):
+        """
+        Register a host as the datatrust in protocol
+        """
+        log.info('****Registering host****')
+        register = send(self.w3, self.datatrust_key, self.datatrust.register(self.datatrust_host, {'from': self.datatrust_wallet}))
+        voting = Voting(self.datatrust_wallet)
+        voting.at(self.w3, self.voting_contract)
+        datatrust_hash = self.w3.sha3(text=self.datatrust_host)
+        is_registered = call(voting.is_candidate(datatrust_hash))
+        if is_registered:
+            log.info(f'Backend registered with transaction {is_registered}. Voting is now open.')
+        else:
+            log.error('Host attempted to register but did not succeed')
+
+    def send_data_hash(self, listing, data_hash):
+        """
+        On a successful post to the API db, send the data hash to protocol
+        """
+        receipt = send(self.w3, self.datatrust_key, self.datatrust.set_data_hash(listing, data_hash, {'from': self.datatrust_wallet}))
```

Comment:
don't need to set from if you instantiated the wallet with that address

Reply:
if thats the case you don't need to pass anything (i.e don't pass {})

Reply:
Updated